app/server/controllers/login.ts

```diff
@@ -25,54 +45,74 @@ function getLogout(req: Request, res: Response) {
   });
 }
 
-function postLogin(getOnlineHearing) {
-  return async(req: Request, res: Response, next: NextFunction) => {
-    const email = req.body['email-address'];
-    const validationMessage = loginEmailAddressValidation(email);
-    if (validationMessage) {
-      const emailAddress = {
-        value: email,
-        error: validationMessage
-      };
-      return res.render('login.html', { emailAddress });
+function getIdamCallback(
+  getToken: (code: string, protocol: string, hostname: string) => TokenResponse,
+  getUserDetails: (accessToken: string) => UserDetails,
+  getOnlineHearing: (email:string) => superAgent.Response,
+  getRedirectUrl: (protocol: string, hostname: string) => string) {
+  return async (req: Request, res: Response, next: NextFunction) => {
+
+    const code: string = req.query.code;
+
+    console.log("Code: " + code);
+
+    if (!code) {
+      return getLogin(getRedirectUrl, '/login')(req, res);
     }
+
     try {
-      const response = await getOnlineHearing(email);
-      if (response.status === NOT_FOUND || response.status === UNPROCESSABLE_ENTITY) {
-        logger.info(`Know issue trying to find hearing for ${email}, status ${response.status}`);
-        const emailAddress = {
-          value: email,
-          error: i18n.login.emailAddress.error[`error${response.status}`]
-        };
-        return res.render('login.html', { emailAddress });
-      }
-      const hearing: OnlineHearing = response.body;
-      req.session.hearing = hearing;
-      logger.info(`Logging in ${email}`);
-      if (hearing.decision && hearing.decision.decision_state === CONST.DECISION_ISSUED_STATE) {
-        logger.info(`Decision issued for hearing with ID ${hearing.online_hearing_id}`);
-        return res.redirect(Paths.decision);
-      }
-      return res.redirect(Paths.taskList);
+      const tokenResponse: TokenResponse = await getToken(code, req.protocol, req.hostname);
+      const userDetails: UserDetails = await getUserDetails(tokenResponse.access_token);
+      // todo Maybe need to check userDetails.accountStatus is 'active' and userDetails.roles contains 'citizen' on userDetails
+
+      const email: string = userDetails.email;
+
+      return await loadHearingAndEnterService(getOnlineHearing, email, req, res)
     } catch (error) {
       AppInsights.trackException(error);
       return next(error);
     }
   };
 }
 
+async function loadHearingAndEnterService(
```

Comment:
This doesn't feel quite right living in this controller. Especially as it's also used by another controller. Can we split up the functionality into `loadHearing` and the rest. It's half a bit of middleware and half a service.

Don't mind if you want to do this in a separate PR to keep it cleaner.

Reply:
Agreed will do this in a separate PR please.

Reply:
Yep that's cool, will be much easier to review also.